cogito/put.go

```diff
@@ -26,91 +45,167 @@ import (
 // Additionally, the script may emit metadata as a list of key-value pairs. This data is
 // intended for public consumption and will make it upstream, intended to be shown on the
 // build's page.
-func Put(log hclog.Logger, in io.Reader, out io.Writer, args []string) error {
-	var pi PutInput
+func Put(log hclog.Logger, in io.Reader, out io.Writer, args []string, putter Putter) error {
+	if err := putter.LoadConfiguration(in, args); err != nil {
+		return fmt.Errorf("put: %s", err)
+	}
+
+	if err := putter.ProcessInputDir(); err != nil {
+		return fmt.Errorf("put: %s", err)
+	}
+
+	// We invoke all the sinks and keep going also if some of them return an error.
+	var sinkErrors []error
+	for _, sink := range putter.Sinks() {
+		if err := sink.Send(); err != nil {
+			sinkErrors = append(sinkErrors, err)
+		}
+	}
+	if len(sinkErrors) > 0 {
+		return fmt.Errorf("put: %s", multiErrString(sinkErrors))
+	}
+
+	if err := putter.Output(out); err != nil {
+		return fmt.Errorf("put: %s", err)
+	}
+
+	return nil
+}
+
+// ProdPutter is an implementation of a [Putter] for the Cogito resource.
+// Use [NewPutter] to create an instance.
+type ProdPutter struct {
+	ghAPI  string
+	log    hclog.Logger
+	gitRef string
+
+	Pi       PutInput
+	InputDir string
+}
+
+// NewPutter returns a Cogito ProdPutter.
+func NewPutter(ghAPI string, log hclog.Logger) *ProdPutter {
+	return &ProdPutter{
+		ghAPI: ghAPI,
+		log:   log,
+	}
+}
+
+func (pu *ProdPutter) LoadConfiguration(in io.Reader, args []string) error {
 	dec := json.NewDecoder(in)
 	dec.DisallowUnknownFields()
-	if err := dec.Decode(&pi); err != nil {
+	if err := dec.Decode(&pu.Pi); err != nil {
 		return fmt.Errorf("put: parsing JSON from stdin: %s", err)
 	}
-	pi.Env.Fill()
+	pu.Pi.Env.Fill()
 
-	if err := pi.Source.ValidateLog(); err != nil {
+	if err := pu.Pi.Source.ValidateLog(); err != nil {
 		return fmt.Errorf("put: %s", err)
 	}
-	log = log.Named("put")
-	log.SetLevel(hclog.LevelFromString(pi.Source.LogLevel))
+	pu.log = pu.log.Named("put")
+	pu.log.SetLevel(hclog.LevelFromString(pu.Pi.Source.LogLevel))
 
-	log.Debug("started",
-		"source", pi.Source,
-		"params", pi.Params,
-		"environment", pi.Env,
+	pu.log.Debug("started",
+		"source", pu.Pi.Source,
+		"params", pu.Pi.Params,
+		"environment", pu.Pi.Env,
 		"args", args)
 
-	if err := pi.Source.Validate(); err != nil {
+	if err := pu.Pi.Source.Validate(); err != nil {
 		return fmt.Errorf("put: %s", err)
 	}
 
 	// args[0] contains the path to a directory containing all the "put inputs".
 	if len(args) == 0 {
```

Comment:
For my understanding, should this check moved on the top?

Reply:
I think you mean "should this check be moved at the beginning of the function?".
I considered this, and I am not even 100% sure that what I did is the best approach. The reason it is this way has been mentioned in a previous PR I think: it is to be able to know the log level :-(

The log level is passed too late, as part of the "source" configuration. This is why "LoadConfiguration" is a mess: it needs to keep going until it gets the log level, so that it can give as much diagnostic information as possible.

If I were to put this check at the beginning of the function, and the check failed, and the log level where set to debug, we could not print any log, only the error message....

One might ask: why we didn't have this problem before, with "ofcourse" ?
Because ofcourse somehow "cheated": the log level was part of ofcourse, and it was ofcourse that was doing the JSON parsing, and calling Put() with a properly initialized logger.

So one might ask: so what is the advantage of removing ofcourse?
The advantage is type safety and LESS code. Ofcourse was obliged to parse the JSON input as a typeless map string->string (as a Python implementation would do) and leave the input validation (both types and field names) to the implementation (you could compare the current master of cogito). On the other hand, this series of PR is doing it the Go idiomatic way: type safety AND field names are ensured automatically by the fact that we use Go structs with a 1-to-1 mapping to the resource YAML configuration (if you search in this series of PRs for JSON parsing tests, you will see that sometimes I pass raw text with some JSON objects, and the input is refused automatically by the JSON parser, without even reaching the cogito code).

Reply:
Thanks for taking the time it's much clearer now.